### PR TITLE
random tests for (a subset of) regular expressions

### PR DESCRIPTION
**Pull request by cfbolz**
Generate random regular expressions, plus matching strings, then make sure they
match. Also try to generate a definitely non-matching string and make
sure it doesn't match (the latter only works in a subset of the cases).

This would have found a pretty bad clunker of a bug in PyPy's re engine: https://foss.heptapod.net/pypy/pypy/commit/c83c263f9f00d18d48ef536947c9b61ca53e01a2

I realize this is quite a lot of non-trivial code, if you don't feel like reviewing this all I'd understand!

### CODE REVIEW

**Comment by cfbolz** [comment]
Ah, I have some `%` formatting left, will fix in a bit.

File: tests/test_re.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+            base(draw)
+            for i in range(draw(st.integers(min_value=2, max_value=MAXREPEAT)))
+        ]
+        if IS_PYPY and sys.pypy_version_info < (7, 3, 1):
+            # PyPy before 7.3.1 actually has broken group references!
+            return Sequence(bases)
+        # the following code would have found the bug in
+        # https://foss.heptapod.net/pypy/pypy/commit/c83c263f9f00d18d48ef536947c9b61ca53e01a2
+        group = draw(st.integers(min_value=0, max_value=len(bases) - 1))
+        name = gensym()
```
Comment (Zac-HD) [inline tests/test_re.py]: You can use the [`st.shared()` strategy](https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.shared) to manage shared state, like:

```python
used_names = draw(st.shared(st.builds(set), key="group names"))  # generates then caches a set
name = draw(st.text().filter(lambda s: s not in used_names)  # or integers().map("symbol{}".format), etc
used_names.add(name)
```

And tada, no more global state so examples will be truly reproducible.

File: tests/test_re.py [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+                res.append(f"{x[0]}-{x[1]}")
+            else:
+                res.append(x)
+        return "[" + "".join(res) + "]"
+
+    @staticmethod
+    def make(draw):
+        # XXX character classes
+        elements = []
+        for _ in range(draw(st.integers(min_value=2, max_value=20))):
```
Comment (Zac-HD) [inline tests/test_re.py]: This is a really tempting pattern, but for internal performance (both speed and quality) reasons it's best to use the `lists()` strategy.  For example:

```python
char = st.characters(blacklist_characters="-^]\\")
elem = char | st.tuples(char, char).filter(lambda t: t[0] != t[1]).map(sorted).map(tuple)
return Charset(draw(st.lists(elem, min_size=2, max_size=20))
```

Admittedly the `elem` definition is a little clumsy; you might perfer to use a `@st.composite` strategy for that bit.

**Review by Zac-HD — COMMENTED** [review]
This generally looks good - thanks!  Ideally I'll do a proper review and merge this in time for the Language Summit, but we'll see.  Preliminary comments below :smile: 

My main request is for more comments and docstrings explaining the general approach you're taking, so that it's clear just from reading the file in isolation.  (it makes sense to me, but probably only because I spent a few days messing with `greenry` recently)  

I'd also prefer names that imply this is a *subset* of the `re` tests, because I'll still want to use other approaches too (e.g. compare Python and C implementations) - maybe `test_re_by_construction`?

**Comment by cfbolz** [comment]
Thanks for the review, and thanks for telling me about greenery! Should get to it soon, hopefully.

File: tests/test_re.py (hunk 1 above)
Comment (cfbolz) [reply tests/test_re.py]: fixed this, thanks for teaching me this feature!

File: tests/test_re.py (hunk 2 above)
Comment (cfbolz) [reply tests/test_re.py]: makes sense! I used a composite indeed

**Comment by cfbolz** [comment]
I resolved your preliminary comments, will try to write some more general text about how the approach works.

Btw, I saw that hypothesmith also doesn't generate unicode identifiers, it's not entirely trivial, is it?

**Comment by Zac-HD** [comment]
It looks like I've specifically overridden this approach for Hypothesmith - presumably it broke something, I'll investigate and either change or annotate it - but the natural approach to generate identifiers is probably `st.from_regex("\w+", fullmatch=True).filter(str.isidentifier)`.

**Comment by cfbolz** [comment]
Ok, I've at least added docstrings everywhere, hope that makes it a bit less impenetrable.

(`st.from_regex("\w+", fullmatch=True).filter(str.isidentifier)` doesn't seem to work, since too few matching things are generated. Anyway, that's not the most important extension to this file 😉)

**Comment by Zac-HD** [comment]
Yep, that's great - thanks again!  FYI I've made a few further tweaks:

- Added you to the contributors list :heart_eyes: 
- Added an assertion that `can_match_empty` is consistent with actual behaviour
- Tweaked the perf check to use a context manager and `perf_counter()` instead of `time()`
- Added a magic attribute to ensure that each logical test has a separate DB entry.  This is Hypothesis internal magic and rarely even useful, but nice for the new fuzz workflow I want to use here :wink:

**Comment by cfbolz** [comment]
Awesome, thank you! What's the fuzz workflow? sounds exciting!

**Comment by cfbolz** [comment]
Btw, I've seen a hanging test now twice, once in CI and once on my laptop. I think it was `test_disassembly`, but can't be sure. I'm a bit confused which part of that test can take arbitrarily long.

**Comment by Zac-HD** [comment]
[The recent fuzzing support](https://hypothesis.readthedocs.io/en/latest/details.html#fuzz-one-input) makes it really easy to turn Hypothesis tests into targets for python-afl or other coverage guided fuzzers. You just call it with a bytestring, see what happens, and repeat! Particularly satisfying aspects include the clean integration with our usual replay and shrinking features, and how easy it can be to generate (and optionally distil) an initial corpus.

Re the hanging test, I've seen it happen occasionally but not tracked it down yet... maybe something about certain recursive structures since an internal change a while back? It's on the Hypothesis issue tracker but nobody's gotten to it yet, if you have any insights or a (even probabilistically) reproducing case that would help a lot.

**Pull request merged**